compiler/rustc_middle/src/mir/interpret/value.rs

```diff
@@ -110,7 +111,8 @@ impl<'tcx> ConstValue<'tcx> {
 /// These variants would be private if there was a convenient way to achieve that in Rust.
 /// Do *not* match on a `Scalar`! Use the various `to_*` methods instead.
 #[derive(Clone, Copy, Eq, PartialEq, TyEncodable, TyDecodable, Hash)]
-#[derive(HashStable)]
+#[derive(HashStable, TypeFoldable, TypeVisitable)]
+#[no_traverse]
```

Comment:
When defining such a type, how do I know whether I have to add this attribute or not? Usually I just copy-paste `derive` clauses from somewhere until things build (because the involves types mean basically nothing to me here). Will that also work for `no_traverse`?

Reply:
The derive macros themselves have some documentation for the `#[no_traverse]` attribute:

> If the annotated type is further annotated with `#[no_traverse]` then the fold will
> merely be a no-op that returns `self`. Otherwise, the fold will construct a value of
> the same struct or enum variant as the input, with each field respectively folded using
> the `TypeFoldable` implementation for its type. However, if a field of a struct or an
> enum variant is itself annotated with `#[no_traverse]`, then that field will retain its
> incumbent value (and its type is not required to implement `TypeFoldable`).

(and similar documentation for `TypeVisitable`).  Let me know if it needs further clarification.

Reply:
Indeed some existing implementations may have been created needlessly, as you say purely in order to satisfy the compiler, and by applying `#[no_traverse]` on the original fields those implementations could now be pruned.

Reply:
Ah, I hadn't realized derive macros can have documentation....

That documentation doesn't tell me when I should use that attribute. If I use it incorrectly, what's the worst that can happen -- unsoundness? Incremental ICEs? Or just bad perf?

Reply:
As I understand it...

Folders and visitors can only take interest in a few key types (Binder, Ty, Region, Const and Predicate).

`TypeFoldable` and `TypeVisitable` only need be implemented for types that might (transitively) contain such "interesting" types.  Any of their fields that *don't* (transitively) contain such "interesting" can be marked `#[no_traverse]` to avoid needlessly descending into them in search of a type of interest.

I think for historical reasons (namely ease of implementation by merely deriving `TypeFoldable` and `TypeVisitable` before uninteresting fields could be skipped with `#[no_traverse]`), implementations of `TypeFoldable` and `TypeVisitable` were required on types that can never contain such "interesting" types.  These implementations were essentially no-ops, and can now be replicated by annotating the type itself with `#[no_traverse]`.

If that understanding is correct, then ultimately any type that is annotated `#[no_traverse]` shouldn't really implement the traits at all but rather should be skipped where they appear in containing types.

As regards what might happen if `#[no_traverse]` is used incorrectly: if it's not applied where it could be, further types may needlessly be required to implement the traits (with potential perf impact); if it's applied where it shouldn't be, possible ICEs or even miscompilation could occur but it shouldn't give rise to any unsoundness within the compiler.

Reply:
By "unsoundess" i meant unsoundess in the resulting code, so sounds like that is on the table. That makes this a rather scary attribute. Shouldn't every use of it be justified? Right now, if someone were to add a `Ty` field to some `no_traverse` type, I can almost guarantee that they would not remember to remove the attribute.

I think the reason that TypeFoldable was implemented by recursion for all types was to make it impossible to forget to handle any nested Ty that might appear anywhere.

Reply:
It certainly isn't impossible to forget to handle nested types of interest even before this PR: the `TrivialTypeTraversalImpl` macro creates a no-op traversal irrespective of the type's fields; decorating a type with `#[no_traverse]` merely replicates that existing behaviour—a `Ty` field added to the type will similarly be skipped in both cases.  Even explicit implementations might not get updated when new fields are added to the type.

However I do see your point, and agree there's danger here.  I'll have a think about how best to address it.  I'm leaning toward removing the ability to decorate a type `#[no_traverse]` and only permitting it on fields—perhaps under another, more clearly dangerous name?  That could at least reduce (though not eliminate) the danger, and bring it on a par with the status quo?

Reply:
Actually it's probably better if types are decorated rather than fields, so that if a type were to change then it only need be updated in one place.

Reply:
Renaming the attribute `#[trivial_traversal]` would keep the language consistent with the status quo, but perhaps something like `#[contains_nothing_interesting_for_traversers]` is more explicit?

I don't know whether that's sufficient to mitigate the danger, but is it objectively any worse than the status quo, where such a trivial impl can be added anyway (using the macro or by hand)?

Indeed, IMHO it's an improvement, as the attribute is on the type definition where it is more likely to be seen and removed should the definition change; rather than hidden in (often) another module of (in many cases\*) another crate.

\*actually, in these cases (types defined upstream of `rustc_middle`), is it not indeed impossible for the type ever to contain a type of interest, as that would entail a circularity in the crate dependency graph?  At least impossible without moving the type downstream into `rustc_middle` or the type of interest upstream?

I'm wondering if this warrants an MCP.

Reply:
I didn't know about `TrivialTypeTraversalImpl`, maybe it is more recent than my last interactions with the traversable traits.

Is there any chance that the `contains_nothing_interesting_for_traversers` attribute could check that indeed none of the field types implements the traversable trait? I can't see how that could be possible but maybe there is a trick...

Either way :+1: for a more clear name. Not sure about MCP, since I'm not in t-compiler, but seeing the other MCPs there this might be good -- and gives a chance for more people to provide feedback on how to design the derive macros here.

Reply:
As it happens, I actually looked into the history earlier today and found that the `TrivialTypeTraversalImpl` macro began life as `CloneTypeFoldableImpls`, introduced by Niko in [`23837c1`](https://github.com/rust-lang/rust/commit/23837c1901ab234ee6bc806744cd47d5a652cdfc#diff-8429aacedec61768fef121af0c31888aedabfd15c751fe03b9a8225b4bffdabcR169-R201) around five years ago.

The attribute could lexically compare types named in each field definition against a blacklist, but that won't catch type aliases or renamed imports or unlisted types that themselves contain types of interest.  So it could add some guard against obvious errors, but would not be reliable.  I'm not aware of any trick by which a procmacro can get actual type information from the compiler (it is expanded before such analysis, after all).

I'll draft an MCP tomorrow.  Thank you for all your thoughts and suggestions!

Reply:
Thanks to @lcnr I've now constrained the applicability of this attribute to fields/types that impl an auto-trait (for which the "types of interest" have negative impls).  This ensures that it cannot be applied incorrectly and therefore I've opted not to give it such a verbose name (though I'm more than happy to go with something else if it's still preferred): `#[skip_traversal]`.

I'll mark this conversation as resolved for now, but by all means just shout if there are still concerns.